scripts/ASTImportTest.sh

```diff
@@ -131,6 +146,125 @@ function test_ast_import_export_equivalence
     TESTED=$((TESTED + 1))
 }
 
+function test_evmjson_import_export_equivalence
+{
+    local sol_file="$1"
+    local input_files=( "${@:2}" )
+    local outputs=( "asm" "bin" "bin-runtime" "opcodes" "srcmap" "srcmap-runtime" )
+    local export_command=("$SOLC" --combined-json "$(IFS=, ; echo "${outputs[*]}")" --pretty-json --json-indent 4 "${input_files[@]}")
+    local success=1
+    if ! "${export_command[@]}" > expected.json 2> expected.error
+    then
+        success=0
+        printError "ERROR: (export) EVM Assembly JSON reimport failed for ${sol_file}"
+        if [[ $EXIT_ON_ERROR == 1 ]]
+        then
+            print_used_commands "$(pwd)" "${export_command[*]}" ""
+            return 1
+        fi
+    fi
+    if ! "${export_command[@]}" "--optimize" > expected.optimize.json 2> expected.optimize.error
+    then
+        success=0
+        printError "ERROR: (export with --optimize) EVM Assembly JSON reimport failed for ${sol_file}"
+        if [[ $EXIT_ON_ERROR == 1 ]]
+        then
+            print_used_commands "$(pwd)" "${export_command[*]} --optimize" ""
+            return 1
+        fi
+    fi
+
+    for input_json in "expected.json" "expected.optimize.json"
+    do
+        local optimize_param=""
+        if [[ "$input_json" == "expected.optimize.json" ]]
+        then
+            optimize_param="--optimize"
+        fi
+
+        # Note that we have some test files, that only consists of free functions.
+        # Those files doesn't define any contracts, so the resulting json does not define any
+        # keys. In this case `jq` returns an error like `jq: error: null (null) has no keys`
+        # to not get spammed by these errors, errors are redirected to /dev/null.
+        for contract in $(jq '.contracts | keys | .[]' $input_json 2> /dev/null)
+        do
+            for output in "${outputs[@]}"
+            do
+                jq --raw-output ".contracts.${contract}.\"${output}\"" "$input_json" > "expected.${output}.json"
+            done
+
+            assembly=$(cat expected.asm.json)
+            [[ $assembly != "" && $assembly != "null" ]] || continue
+
+            local import_command=("${SOLC}" --combined-json "bin,bin-runtime,opcodes,asm,srcmap,srcmap-runtime" --pretty-json --json-indent 4 --import-asm-json expected.asm.json)
+            if ! "${import_command[@]}" > obtained.json 2> obtained.error
```

Comment:
Shouldn't `import_command` also use `$optimize_param` to match the export command? Otherwise I can't see how the bytecode from both could match.

Reply:
basically, for the import it is not important to know whether the imported assembly was initially optimised or not. It just takes the given input and imports it. I was wondering whether it really makes sense to have that additional import of the optimised byte code. but I thought that it would make sense to test the import on as many different inputs as possible. I only have `$optimize_param` here, so that the user is able to know exactly how the used input was generated, so that a specific test can be reproduced.